src/LinuxTracing/LeafFunctionCallManager.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2021 The Orbit Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "LeafFunctionCallManager.h"
+
+#include <sys/mman.h>
+
+#include <vector>
+
+#include "OrbitBase/Logging.h"
+
+namespace orbit_linux_tracing {
+
+using orbit_grpc_protos::Callstack;
+
+// Let's unwind the stack from $rsp to $rbp using libunwindstack. If $rbp points to the current
+// frame, this will only include the locals and not the return address, so libunwindstack will
+// only be able to unwind one frame (the instruction pointer). If $rpb points to the previous frame,
+// as the leaf function does not contain frame pointers, the region from $rsp to $rbp will contain
+// the complete innermost frame, plus the locals of the caller's frame (but not the return address).
+// Therefore, libunwindstack will be able to compute two frames, where the outer one is the missing
+// caller and needs to be patched in.
+// As libunwindstack will try to unwind even further, unwinding errors will always be reported.
+// We need to fallback to plausibility checks to detect actual unwinding errors (such as whether the
+// frames are executable).
+// Note that we cannot simply set libunwindstack to unwind always two frames and compare the outer
+// frame with the respective one in the callchain carried by the perf_event_open event, as in case
+// of uprobes overriding the return addresses, both addresses would be identical even if the actual
+// addresses (after uprobe patching) are not.
+// More (internal) documentation on this in: go/stadia-orbit-leaf-frame-pointer
+Callstack::CallstackType LeafFunctionCallManager::PatchLeafFunctionCaller(
+    CallchainSamplePerfEvent* event, LibunwindstackMaps* current_maps,
+    LibunwindstackUnwinder* unwinder) {
+  CHECK(event != nullptr);
+  CHECK(current_maps != nullptr);
+  CHECK(unwinder != nullptr);
+  CHECK(event->GetCallchainSize() > 2);
+
+  uint64_t stack_size =
+      event->GetRegisters()[PERF_REG_X86_BP] - event->GetRegisters()[PERF_REG_X86_SP];
+  if (stack_size > SAMPLE_STACK_USER_SIZE_512BYTES) {
+    return Callstack::kFramePointerDwarfStackTooSmallError;
+  }
+
+  const LibunwindstackResult& libunwindstack_result =
+      unwinder->Unwind(event->GetPid(), current_maps->Get(), event->GetRegisters(),
+                       event->GetStackData(), stack_size, true);
+  const std::vector<unwindstack::FrameData>& libunwindstack_callstack =
+      libunwindstack_result.frames();
+
+  if (libunwindstack_callstack.empty()) {
+    ERROR("Discarding sample as DWARF-based unwinding resulted in empty callstack (error: %s)",
+          LibunwindstackUnwinder::LibunwindstackErrorString(libunwindstack_result.error_code()));
+    return Callstack::kDwarfUnwindingError;
```

Comment:
This is the type that is then sent with the incomplete callstack (correct me if I'm wrong). I was wondering whether it would be better the have a separate type for dwarf unwinding errors that happen here (e.g., `kStackTopDwarfUnwindingError`), to distinguish between these and the ones that happen when actually using full DWARF-unwinding? Did you think about this or what do you think?

Reply:
Would be fine for me. My intend was that it is actually the same kind of error: We do Libunwindstack and that fails (for whatever internal reason).

Reply:
Added a `kStackTopDwarfUnwindingError`